content/en/tracing/profiler/intro_to_profiling.md

```diff
@@ -0,0 +1,193 @@
+---
+title: Intro to Profiling
+kind: Documentation
+---
+
+You may have heard that profiling can make your services faster, cheaper, and more reliable, but if you haven't used a profiler before, it can seem like a dark art.
+
+In this guide, we'll explain profiling, then profile a sample service that has a performance problem and see how we can use Continuous Profiler to understand and fix the problem!
+
+## What is Profiling?
+
+A profiler shows how much "work" each line of code is doing by collecting data about the program as it's running. For example, infrastructure monitoring might show that your app servers are using 80% of their CPU but you may not know why. Profiling could then show you that function `doSomeWork` is using 48% of the CPU, function `renderGraph` is using another 19%, and so on. This is important when working on performance problems because many programs spend a lot of time in very few places and it's often not obvious which places those are. Guessing at which parts of a program to optimize can often cause engineers to spend a lot of time and not get much in the way of results. By using a profiler, we find exactly which parts of the code we should optimize to get the most bang for our buck.
+
+If you've used an APM tool, you might think of profiling like a "deeper" tracer that provides a very fine grained view of your code without needing any instrumentation.
+
+Datadog Continuous Profiler can track various types of "work", including CPU usage, amount and types of objects being allocated in memory, time spent waiting to acquire locks, amount of network or file I/O, and more! The profile types available depend on the language being profiled.
+
+We've got an [example service][1] with a performance problem that you can easily experiment with. The example service has an API which allows searching a "database" of 5000 movies. We'll fix a performance problem with it! This guide shows you the process, but for the full experience you can follow along with your own shell, browser, and IDE.
+
+## Prerequisites
+
+You'll need:
+1. [docker-compose][2]
+2. A Datadog account and [API key][3] (you don't need an application key). If you don't already have a Datadog account, [sign up for a free trial][4].
+
+## Run the Example
+
+Get the example service up and running with:
+```
+git clone https://github.com/DataDog/dd-continuous-profiler-example.git
+cd dd-continuous-profiler-example
+echo "DD_API_KEY=YOUR_API_KEY" > docker.env
+docker-compose up -d
```

Comment:
I just tried running this and I'm getting `apt-get` errors. The error message mentions `Mirror sync in progress?`, so this might be intermittent. But I figured I should mention it.

https://gist.github.com/felixge/5088c032e7f395ab9f4affb69d1e4a5b

Reply:
Thanks for flagging! I expect you're right that it's a transient issue with whatever apt mirror is being used. It's working for me now.

Reply:
I just ran it again and I'm still seeing this error. I also tried `docker-compose build --no-cache` to see if that would fix it, but same result.

Maybe you can try `docker-compose build --no-cache` on your end as well to compare?

Or do you have a better idea for cleaning the build cache on my end?

Reply:
Weird. I tried it on a new machine and it worked. I also just tried running apt-get update with security.debian.org set to each of the ips it resolves to for me via /etc/hosts and they all worked.

If you're still seeing this, what does that hostname resolve to?

Reply:
> Weird. I tried it on a new machine and it worked. I also just tried running apt-get update with security.debian.org set to each of the ips it resolves to for me via /etc/hosts and they all worked.
> If you're still seeing this, what does that hostname resolve to?

How did you modify the `/etc/hosts` file?

I ran `dig security.debian.org` on my host and tried to put the first IP `151.101.0.204` into the /etc/hosts like this:

```diff
diff --git a/Dockerfile.toolbox b/Dockerfile.toolbox
index 26e2b7f..2807459 100644
--- a/Dockerfile.toolbox
+++ b/Dockerfile.toolbox
@@ -1,4 +1,6 @@
 FROM httpd:2.4
 
+RUN echo '151.101.0.204' >> /etc/hosts
+RUN cat /etc/hosts
 RUN apt-get -y update && apt-get install -y curl jq
 CMD tail -f /dev/null
```

Unfortunately docker seems to aggressively control modifications to this file, so it seems like my change gets immediately reverted:

```
Step 2/5 : RUN echo '151.101.0.204' >> /etc/hosts
 ---> Running in 1613ebca94d3
Removing intermediate container 1613ebca94d3
 ---> c1a345676fe9
Step 3/5 : RUN cat /etc/hosts
 ---> Running in db1ba3f2f934
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.2	db1ba3f2f934
```

Same story with `/etc/resolv.conf`.

I can't even figure out what IP `security.debian.org` resolves to inside of the container because there is no `ping`, `dig`, etc. and I can't use `apt-get` to install the tools because of the errors I'm getting ... 🤕 

Use containers they said, it's going to be great the said 😂.

![](https://media.giphy.com/media/bAftZ12SC0uEjLndIh/giphy.gif)